src/v/compression/internal/gzip_compressor.cc

```diff
@@ -117,106 +117,114 @@ class gzip_decompression_codec {
-    const char* _input;
-    size_t _input_size;
+
+    const iobuf& _input;
+    iobuf::const_iterator _input_chunk;
+
     gz_header _hdr; // needed for gzip
     z_stream _stream;
 };
 
 iobuf gzip_compressor::compress(const iobuf& b) {
-    ss::temporary_buffer<char> obuf;
-    {
-        gzip_compression_codec def;
-        def.reset();
-        z_stream& strm = def.stream();
-        /* Calculate maximum compressed size and
-         * allocate an output buffer accordingly, being
-         * prefixed with the Message header. */
-        const size_t output_size = deflateBound(&strm, b.size_bytes());
-        obuf = ss::temporary_buffer<char>(output_size);
+    const size_t max_chunk_size = details::io_allocation_size::max_chunk_size;
 
-        // NOLINTNEXTLINE
-        strm.next_out = (unsigned char*)obuf.get_write();
-        strm.avail_out = output_size;
-
-        /* Iterate through each segment and compress it. */
-        for (auto& io : b) {
-            // zlib is not const correct
-            // NOLINTNEXTLINE
-            strm.next_in = (unsigned char*)io.get();
-            strm.avail_in = io.size();
-            throw_if_zstream_error(
-              "gzip error compressing chunk: {}", deflate(&strm, Z_NO_FLUSH));
+    gzip_compression_codec def;
+    def.reset();
+    z_stream& strm = def.stream();
+
+    const size_t output_size_estimate = deflateBound(&strm, b.size_bytes());
+    size_t output_chunk_size = std::min(max_chunk_size, output_size_estimate);
+
+    ss::temporary_buffer<char> obuf(output_chunk_size);
+    strm.next_out = reinterpret_cast<unsigned char*>(obuf.get_write());
+    strm.avail_out = obuf.size();
+
+    iobuf ret;
+    auto frag_i = b.begin();
+    while (true) {
+        if (strm.avail_in == 0 && frag_i != b.end()) {
+            strm.next_in = const_cast<unsigned char*>(
+              reinterpret_cast<const unsigned char*>(frag_i->get()));
+            strm.avail_in = frag_i->size();
+            frag_i++;
         }
-        /* Finish the compression */
-        if (int ret = deflate(&strm, Z_FINISH); ret != Z_STREAM_END) {
-            throw_if_zstream_error("gzip error finishing compression: {}", ret);
+
+        if (strm.avail_out == 0) {
+            obuf.trim(output_chunk_size - strm.avail_out);
+            ret.append(std::move(obuf));
+            output_chunk_size = std::min(output_chunk_size * 2, max_chunk_size);
+            obuf = ss::temporary_buffer<char>(output_chunk_size);
+            strm.next_out = reinterpret_cast<unsigned char*>(obuf.get_write());
+            strm.avail_out = obuf.size();
         }
-        obuf.trim(def.stream().total_out);
-        // trigger deflateEnd
+
+        int flush = (strm.avail_in == 0) ? Z_FINISH : Z_NO_FLUSH;
+        int deflate_r = deflate(&strm, flush);
+        if (deflate_r == Z_STREAM_END) {
+            break;
+        }
+        throw_if_zstream_error(
+          "gzip error compressing chunk: {} (flush={})", deflate_r);
```

Comment:
mismatched number of format placeholders and arguments?

Reply:
oops, fixed.